src/lib/oper_data.c

```diff
@@ -182,37 +179,279 @@ int get_list_keys(const struct lysc_node_list *list, json_object *json_array_obj
-                return EXIT_FAILURE;
+                ret = EXIT_FAILURE;
+                goto cleanup;
             }
         }
     }
     (*keys_num) = key_count;
-    return EXIT_SUCCESS;
+    ret = EXIT_SUCCESS;
+cleanup:
+    free(key_name);
+    return ret;
+}
+
+/**
+ * Converts a semicolon-separated key-value string to a JSON object. Each pair is separated by ';', and key-value within a pair is separated by ':'.
+ * This function is useful for converting OPER_VALUE_MAP_EXT and OPER_FLAG_MAP_EXT mapping from string to JSON objects for easier manipulation and access.
+ * @param [in] input: A semicolon-separated key-value pairs string.
+ * @return A pointer to a JSON object representing the input string's key-value pairs. The caller is responsible for freeing the JSON object.
+ */
+json_object *strmap_to_jsonmap(const char *input)
+{
+    struct json_object *jobj = json_object_new_object();
+    char *pairs = strdup(input);
+    char *pair;
+    char *rest_pairs = pairs;
+
+    while ((pair = strtok_r(rest_pairs, ";", &rest_pairs))) {
+        char *rest_kv = pair;
+        char *key = strtok_r(rest_kv, ":", &rest_kv);
+        char *value = strtok_r(NULL, ":", &rest_kv);
+
+        if (key && value) {
+            json_object_object_add(jobj, key, json_object_new_string(value));
+        }
+    }
+
+    free(pairs);
+    return jobj;
+}
+
+/**
+ * Looks up the value associated with a given key in a JSON object and returns it. If the key isn't found, returns the original value.
+ * 
+ * For a given value (original_value) this function helps mapping it to its corresponding mapped value found in value map JSON Object.
+ * If orignal_value doesn't have a corresponding mapping, it gets returned as is.
+ * @param [in] value_map_jobj: JSON object containing key-value mappings.
+ * @param [in] original_value: The key to look up in the mapping.
+ * @return The mapped value if the key is found; otherwise, the original value is returned.
+ */
+const char *map_value_if_needed(struct json_object *value_map_jobj, const char *original_value)
+{
+    struct json_object *mapped_value_obj;
+    if (value_map_jobj &&
+        json_object_object_get_ex(value_map_jobj, original_value, &mapped_value_obj)) {
+        return json_object_get_string(mapped_value_obj);
+    }
+    return original_value;
+}
+
+/**
+ * This function compairs schema node (s_node) parents names to lyd_node data tree node name, if the 
+ * schema node parent containers are not added to the lyd_node, this function adds them.
+ * 
+ * This function is to be used before processing leafs and leaf_lists to ensure their parent continers
+ * are added to the lyd data tree.
+ * 
+ * @param [in] s_node: schema node being processed (typically the s_node of a leaf or leaf_list).
+ * @param [in] parent_data_node: lyd data tree to which nodes under process are being attached to.
+ */
+void add_parent_node_if_needed(const struct lysc_node *s_node, struct lyd_node **parent_data_node)
```

Comment:
`add_missing_parents` might be a better name,

Reply:
Ok